edx_repo_tools/find_dependencies/find_dependencies.py

```diff
@@ -0,0 +1,335 @@
+        if url:
+            urls.add(canonical_url(url))
+
+    for url in parallel_map(repo_url_from_tgz, Path("files").glob("*.tar.gz"), "Examining tar.gz"):
+        if url:
+            urls.add(canonical_url(url))
+
+    with open("base.txt") as fbase:
+        for line in fbase:
+            if match := re.search(r"https://github.com[^@ #]*(\.git)?", line):
+                urls.add(canonical_url(match[0]))
+
+    real_urls = set()
+    for url in parallel_map(find_real_url, urls, "Getting real URLs"):
+        if url:
+            real_urls.add(url)
+
+    write_list("repo_urls.txt", sorted(real_urls))
+    return real_urls
+
+def matching_text(text, regexes):
+    """Find a line in text matching a regex, and return the first regex group."""
+    for regex in regexes:
+        for line in text.splitlines():
+            if match := re.search(regex, line):
+                return match[1]
+    return None
+
+@cached
+def repo_url_from_wheel(wheel_path: str) -> Optional[str]:
+    """Read metadata from a .whl file, returning the repo URL."""
+    with zipfile.ZipFile(wheel_path) as whl_file:
+        fmetadata = next((f for f in whl_file.namelist() if f.endswith("/METADATA")), None)
+        if fmetadata is None:
+            print(f"No metadata in {wheel_path}")
+            return None
+        with whl_file.open(fmetadata) as inner_file:
+            metadata = inner_file.read().decode("utf-8")
+        return repo_url_from_metadata(wheel_path, metadata)
+
+@cached
+def repo_url_from_tgz(tgz_path: str) -> Optional[str]:
+    """Read metadata from a .tar.gz file, returning the repo URL."""
+    with tarfile.open(tgz_path) as tgz_file:
+        fmetadata = next((f for f in tgz_file.getnames() if f.endswith("/PKG-INFO")), None)
+        if fmetadata is None:
+            print(f"No metadata in {tgz_path}")
+            return None
+        metadata = tgz_file.extractfile(fmetadata).read().decode("utf-8")
+        return repo_url_from_metadata(tgz_path, metadata)
+
+
+SOURCE_URL_REGEXES = [
+    r"(?i)^Project-URL: Source.*,\s*(.*)$",
+    r"(?i)^Home-page: (.*)$",
+    r"(?i)^Project-URL: Home.*,\s*(.*)$",
+]
+
+def repo_url_from_metadata(filename, metadata):
+    """Find the likely source repo URL from PyPI metadata."""
+    repo_url = matching_text(metadata, SOURCE_URL_REGEXES)
+    if repo_url is None:
+        print(f"No repo URL in {filename}")
+        return None
+    if repo_url == "UNKNOWN":
+        print(f"Repo URL is UNKNOWN in {filename}")
+        return None
+    return repo_url
+
+PY_REQS = [
+    "requirements/edx/base.txt",
+    "requirements/base.txt",
+    "requirements.txt",
+]
+
+def find_py_reqs():
+    """Find the Python requirements file to use."""
+    for fname in PY_REQS:
+        possible_req = Path(fname)
+        if possible_req.exists():
+            return possible_req
+    return None
+
+
+def process_directory():
+    """Find all the requirements in the current directory. Returns a set of repo URLs."""
```

Comment:
This function, critically, also copies the js/py requements list to `package-lock.json`/`base.txt`. I would note that in the docstring.